### ISSUE TITLE
JSON-LD Credential Doesn't Validate Credential Attributes are Valid

### ISSUE THREAD

**Issue by ianco** [hyperledger/aries-cloudagent-python#1187]
When submitting a JSON-LD credential, the `@context` defines the set of acceptable attributes for the credential, however this is not validated and any attribute can be provided.

For example in the following, `givenName`, `familyName` and `alumniOf` are defined by `https://schema.org` however `undefinedAttribute` is not defined anywhere.  This credential can be successfully submitted, should this be allowed?

```
{
  "connection_id": "ad35a4d8-c84b-4a4f-a83f-1afbf134b8b9",
  "filter": {
    "ld_proof": {
      "credential": {
        "@context": [
          "https://www.w3.org/2018/credentials/v1",
          "https://schema.org"
        ],
        "type": ["VerifiableCredential", "Person"],
        "issuer": "did:key:zUC71pj2gpDLfcZ9DE1bMtjZGWCSLhkQsUCaKjqXtCftGkz27894pEX9VvGNiFsaV67gqv2TEPQ2aDaDDdTDNp42LfDdK1LaWSBCfzsQEyaiR1zjZm1RtoRu1ZM6v6vz4TiqDgU",
        "issuanceDate": "2020-01-01T12:00:00Z",
        "credentialSubject": {
          "givenName": "Sally",
          "familyName": "Student",
          "alumniOf": "Example University",
          "undefinedAttribute": "Some value",
          "jkashdkajhsdkjahsdkjahsdjkh": "ajskhdakjsdhjaksdh"
        }
      },
      "options": {
        "proofType": "BbsBlsSignature2020"
      }
    }
  }
}
```

**Comment by ianco** [comment]
Likewise, this example also works:

```
{
  "connection_id": "ad35a4d8-c84b-4a4f-a83f-1afbf134b8b9",
  "filter": {
    "ld_proof": {
      "credential": {
        "@context": [
          "https://www.w3.org/2018/credentials/v1",
          "https://schema.org"
        ],
        "type": ["VerifiableCredential", "Person"],
        "issuer": "did:key:zUC71pj2gpDLfcZ9DE1bMtjZGWCSLhkQsUCaKjqXtCftGkz27894pEX9VvGNiFsaV67gqv2TEPQ2aDaDDdTDNp42LfDdK1LaWSBCfzsQEyaiR1zjZm1RtoRu1ZM6v6vz4TiqDgU",
        "issuanceDate": "2020-01-01T12:00:00Z",
        "credentialSubject": {
          "student": {
            "type": "Person",
            "givenName": "Sally",
            "familyName": "Student",
            "alumniOf": "Example University",
            "jakshdkjahsdkasdkahksdhakjshkajhd": "kjashdkajhsdkjahsdkjahdjksh"
          }
        }
      },
      "options": {
        "proofType": "BbsBlsSignature2020"
      }
    }
  }
}
```

**Comment by TimoGlastra** [comment]
This should throw an error

> "\<x> attributes dropped. Provide definitions in context to correct. [\<missing-properties>]"

https://github.com/hyperledger/aries-cloudagent-python/blob/main/JsonLdCredentials.md#json-ld-context

**Comment by TimoGlastra** [comment]
What API endpoint are you calling?

**Comment by ianco** [comment]
> What API endpoint are you calling?

`http://localhost:8021/issue-credential-2.0/send-offer`

**Comment by TimoGlastra** [comment]
For clarification, is the credential already issued in this case? I think it only checks at issuance, but would be good to also check when the API call is made

**Comment by ianco** [comment]
Yes in each case the credential was issued successfully

**Comment by TimoGlastra** [comment]
Ah I was just wondering about that. I ran into the same issue last week. Are you going to merge the changes upstream?

**Comment by ianco** [comment]
> Ah I was just wondering about that. I ran into the same issue last week. Are you going to merge the changes upstream?

Yes I'm working on my own fork right now I'll open a PR later today.

**Comment by TimoGlastra** [comment]
TLDR: don't use `http://schema.org` directly as context.

--- 

Ok did a quick investigation. This is "working as expected".

the context from `schema.org` defines an `@vocab` which sets a default for all properties to extend from.

So the property `jakshdkjahsdkasdkahksdhakjshkajhd` actually becomes `http://schema.org/jakshdkjahsdkasdkahksdhakjshkajhd`.

Most json-ld contexts/vocabularies only use a few specific properties/types from the schema vocab (e.g. https://w3c-ccg.github.io/citizenship-vocab/contexts/citizenship-v1.jsonld).

This jsonld-checker also reports the credential you mentioned as valid: https://www.jsonld-checker.com/?json=%257B%250A%2520%2520%2522%2540context%2522%253A%2520%255B%250A%2520%2520%2520%2520%2522https%253A%252F%252Fwww.w3.org%252F2018%252Fcredentials%252Fv1%2522%252C%250A%2520%2520%2520%2520%2522https%253A%252F%252Fschema.org%2522%250A%2520%2520%255D%252C%250A%2520%2520%2522type%2522%253A%2520%255B%2522VerifiableCredential%2522%252C%2520%2522Person%2522%255D%252C%250A%2520%2520%2522issuer%2522%253A%2520%2522did%253Akey%253AzUC71pj2gpDLfcZ9DE1bMtjZGWCSLhkQsUCaKjqXtCftGkz27894pEX9VvGNiFsaV67gqv2TEPQ2aDaDDdTDNp42LfDdK1LaWSBCfzsQEyaiR1zjZm1RtoRu1ZM6v6vz4TiqDgU%2522%252C%250A%2520%2520%2522issuanceDate%2522%253A%2520%25222020-01-01T12%253A00%253A00Z%2522%252C%250A%2520%2520%2522credentialSubject%2522%253A%2520%257B%250A%2520%2520%2520%2520%2522student%2522%253A%2520%257B%250A%2520%2520%2520%2520%2520%2520%2522type%2522%253A%2520%2522Person%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522givenName%2522%253A%2520%2522Sally%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522familyName%2522%253A%2520%2522Student%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522alumniOf%2522%253A%2520%2522Example%2520University%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522jakshdkjahsdkasdkahksdhakjshkajhd%2522%253A%2520%2522kjashdkajhsdkjahsdkjahdjksh%2522%250A%2520%2520%2520%2520%257D%250A%2520%2520%257D%250A%257D

we can only check for dropped properties and because of the schema.org `@vocab` the properties are not dropped.

**Comment by TimoGlastra** [comment]
If I run the checker in ACA-Py without schema.org it reports it's missing the student property: `['credentialSubject.student']`

**Comment by ianco** [comment]
Interesting thanks for the info @TimoGlastra , I wonder if we should explicitely block this?

Currently we *can* use `https://schema.org` in the context, but then we lose the attribute validation.

**Comment by andrewwhitehead** [comment]
I think it makes sense to block it. If schema.org had a versioned global context then it might make sense to allow that, but the wildcard resolution is a problem. Wonder if `@vocab` can be banned in general?

**Comment by ianco** [comment]
> If I run the checker in ACA-Py without the schema.org context it reports it's missing the student property: `['credentialSubject.student']`

Interestingly the W3C vaccine example fails with similar errors.
